File: library/src/main/java/tw/shounenwind/plurkconnection/BuildablePlurkConnection.java
package tw.shounenwind.plurkconnection;

import android.content.Context;

import java.io.File;
import java.lang.ref.WeakReference;
import java.util.concurrent.ExecutorService;

import tw.shounenwind.plurkconnection.callbacks.ApiNullCallback;
import tw.shounenwind.plurkconnection.callbacks.ApiStringCallback;
import tw.shounenwind.plurkconnection.callbacks.BasePlurkCallback;
import tw.shounenwind.plurkconnection.responses.ApiResponseNull;

public class BuildablePlurkConnection extends PlurkConnection {

    private static final String TAG = "BPC";
    private ExecutorService threadPool;


    public BuildablePlurkConnection(String APP_KEY, String APP_SECRET, String token, String token_secret) {
        super(APP_KEY, APP_SECRET, token, token_secret);

    }

    public BuildablePlurkConnection(String APP_KEY, String APP_SECRET) {
        super(APP_KEY, APP_SECRET);
    }

    public void setThreadPool(ExecutorService threadPool) {
        this.threadPool = threadPool;
    }

    public Builder builder(Context mContext) {
        return new Builder(mContext, this);
    }

    public class Builder {

        private final NewThreadRetryExecutor retryExecutor;
        private WeakReference<Context> contextWeakReference;
        private BuildablePlurkConnection mHealingPlurkConnection;
        private String target;
        private Param[] params;
        private BasePlurkCallback callback;

        public Builder(Context mContext, BuildablePlurkConnection healingPlurkConnection) {
            contextWeakReference = new WeakReference<>(mContext);
            mHealingPlurkConnection = healingPlurkConnection;
            params = new Param[]{};
            retryExecutor = new NewThreadRetryExecutor();
            if (threadPool != null) {
                retryExecutor.setThreadPool(threadPool);
            }
        }

        public Builder setRetryTimes(int retryTimes) {
            retryExecutor.setTotalRetryTimes(retryTimes);
            return this;
        }

        public Builder setTarget(String target) {
            this.target = target;
            return this;
        }

        public Builder setParams(Param[] params) {
            this.params = params;
            return this;
        }

        public Builder setParam(Param param) {
            this.params = new Param[]{param};
            return this;
        }

        public Builder setCallback(BasePlurkCallback callback) {
            this.callback = callback;
            return this;
        }

        public void call() {

            retryExecutor.setTasks(retryExecutor.new Tasks() {
                @Override
                public void mainTask() throws Exception {
                    if (callback == null) {
                        callback = new ApiNullCallback() {
                            @Override
                            public void onSuccess(ApiResponseNull parsedResponse) throws Exception {

                            }
                        };
                    }
                    callback.runResult(mHealingPlurkConnection.startConnect(target, params));
                }

                @Override
                public void onRetry(Throwable e, int retryTimes, int totalTimes) {
                    if (callback != null)
                        callback.onRetry(e, retryTimes, totalTimes, new ErrorAction(this));
                }

                @Override
                public void onError(Throwable e) {
                    if (callback != null)
                        callback.onError(e);
                }
            });
            retryExecutor.run(contextWeakReference.get());

        }

        public void upload(final File imageFile, final String fileName) {
            retryExecutor.setTasks(retryExecutor.new Tasks() {

                public void mainTask() throws Exception {
                    if (!(callback instanceof ApiStringCallback)) {
                        throw new Exception("Callback needs to instanceof ApiStringCallback");
                    }
                    callback.runResult(mHealingPlurkConnection.startConnect(target, imageFile, fileName));
                }

                public void onRetry(Exception e, int retryTimes, int totalTimes) {
                    if (callback != null) {
                        callback.onRetry(e, retryTimes, totalTimes, new ErrorAction(this));
                    }
                }

                public void onError(Exception e) {
                    if (callback != null)
                        callback.onError(e);
                }
            });

            retryExecutor.run(contextWeakReference.get());

        }
    }

    public class ErrorAction {

        private NewThreadRetryExecutor.Tasks tasks;

        public ErrorAction(NewThreadRetryExecutor.Tasks tasks) {
            this.tasks = tasks;
        }

        public final void retry() {
            tasks.retry();
        }

        public final void error(Throwable e) {
            tasks.error(e);
        }
    }

}

File: library/src/main/java/tw/shounenwind/plurkconnection/BuildablePlurkConnection.kt
package tw.shounenwind.plurkconnection

import android.app.Activity
import android.app.Application
import android.content.Context
import android.os.Bundle
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.cancel
import tw.shounenwind.plurkconnection.callbacks.*
import java.io.File
import java.lang.ref.WeakReference

open class BuildablePlurkConnection : PlurkConnection {

    constructor(APP_KEY: String, APP_SECRET: String, token: String, token_secret: String) : super(APP_KEY, APP_SECRET, token, token_secret)

    constructor(APP_KEY: String, APP_SECRET: String) : super(APP_KEY, APP_SECRET)

    fun builder(mContext: Context): Builder {
        return Builder(mContext, this)
    }

    inner class Builder(mContext: Context, private val mHealingPlurkConnection: BuildablePlurkConnection) : Application.ActivityLifecycleCallbacks {
        private val retryExecutor: NewThreadRetryExecutor
        private val contextWeakReference: WeakReference<Context> = WeakReference(mContext)
        private var target: String? = null
        private var params: Array<Param>? = null
        private var callback: BasePlurkCallback<*>? = null
        private var onRetryAction: OnRetryAction? = null
        private var onErrorAction: OnErrorAction? = null
        private var mainScope: CoroutineScope? = MainScope()

        init {
            params = arrayOf()
            retryExecutor = NewThreadRetryExecutor()
        }

        fun setRetryTimes(retryTimes: Int) = apply {
            retryExecutor.setTotalRetryTimes(retryTimes)
        }

        fun setTarget(target: String) = apply {
            this.target = target
        }

        fun setParams(params: Array<Param>) = apply {
            this.params = params
        }

        fun setParam(param: Param) = apply {
            this.params = arrayOf(param)
        }

        fun setCallback(callback: BasePlurkCallback<*>) = apply {
            this.callback = callback
        }

        fun setOnRetryAction(action: OnRetryAction) = apply {
            onRetryAction = action
        }

        inline fun setOnRetryAction(crossinline body: (
                e: Throwable,
                retryTimes: Long,
                totalTimes: Long,
                errorAction: ErrorAction
        ) -> Unit) = apply {
            setOnRetryAction(object: OnRetryAction{
                override fun onRetry(e: Throwable, retryTimes: Long, totalTimes: Long, errorAction: ErrorAction) {
                    body(e, retryTimes, totalTimes, errorAction)
                }

            })
        }

        fun setOnErrorAction(action: OnErrorAction) = apply {
            onErrorAction = action
        }

        inline fun setOnErrorAction(crossinline body: (e: Throwable) -> Unit) = apply {
            setOnErrorAction(object : OnErrorAction{
                override fun onError(e: Throwable) {
                    body(e)
                }
            })
        }

        fun call() {
            val mContext = contextWeakReference.get()?:return
            retryExecutor.setTasks(object : Tasks(mContext) {
                @Throws(Exception::class)
                override fun mainTask() {
                    if (callback == null) {
                        callback = ApiNullCallback()
                    }
                    callback!!.runResult(mHealingPlurkConnection.startConnect(target!!, params!!))
                }

                override fun onRetry(e: Throwable, retryTimes: Int, totalTimes: Int) {
                    onRetryAction?.onRetry(e, retryTimes.toLong(), totalTimes.toLong(), ErrorAction(this))
                }

                override fun onError(e: Throwable) {
                    onErrorAction?.onError(e)
                }
            })
            synchronized(Builder::class.java) {
                retryExecutor.run(mainScope)
            }

        }

        fun upload(imageFile: File, fileName: String) {
            val mContext = contextWeakReference.get()?:return
            retryExecutor.setTasks(object : Tasks(mContext) {

                @Throws(Exception::class)
                override fun mainTask() {
                    if (callback !is ApiStringCallback) {
                        throw Exception("Callback needs to instanceof ApiStringCallback")
                    }
                    callback!!.runResult(mHealingPlurkConnection.startConnect(target!!, imageFile, fileName))
                }

                fun onRetry(e: Exception, retryTimes: Int, totalTimes: Int) {
                    onRetryAction?.onRetry(e, retryTimes.toLong(), totalTimes.toLong(), ErrorAction(this))
                }

                fun onError(e: Exception) {
                    onErrorAction?.onError(e)
                }
            })

            synchronized(Builder::class.java) {
                retryExecutor.run(mainScope)
            }

        }

        override fun onActivityPaused(p0: Activity) {

        }

        override fun onActivityStarted(p0: Activity) {

        }

        override fun onActivityDestroyed(p0: Activity) {
            if (contextWeakReference.get() == p0) {
                synchronized(Builder::class.java) {
                    mainScope?.cancel()
                    mainScope = null
                }
            }
        }

        override fun onActivitySaveInstanceState(p0: Activity, p1: Bundle) {

        }

        override fun onActivityStopped(p0: Activity) {

        }

        override fun onActivityCreated(p0: Activity, p1: Bundle?) {

        }

        override fun onActivityResumed(p0: Activity) {

        }
    }

    inner class ErrorAction(private val tasks: Tasks) {

        fun retry() {
            tasks.retry()
        }

        fun error(e: Throwable) {
            tasks.error(e)
        }
    }

    companion object {
        private const val TAG = "BPC"
    }

}
